File: demo/src/fire.py
#!/usr/bin/env python
# -*- coding: utf-8 -*

# Loosely based on https://github.com/giladaya/arduino-led-matrix/blob/master/fire/fire.ino

from tinkerforge.ip_connection import IPConnection
from tinkerforge.bricklet_led_strip import LEDStrip

import colorsys
import random

import config

class Fire:
    values = [
        [ 32,  16,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,  16,  32],
        [ 64,  32,  16,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,  16,  32,  64],
        [ 96,  64,  32,  32,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,  32,  32,  64,  96],
        [128,  96,  64,  64,  32,  32,   0,   0,   0,   0,   0,   0,   0,   0,  32,  32,  64,  64,  96, 128],
        [160, 128,  96,  96,  64,  64,  32,  32,   0,   0,   0,   0,  32,  32,  64,  64,  96,  96, 128, 160],
        [192, 160, 128, 128,  96,  96,  64,  64,  32,   0,   0,  32,  64,  64,  96,  96, 128, 128, 160, 192],
        [255, 192, 160, 160, 128, 128,  96,  96,  64,  32,  32,  64,  96,  96, 128, 128, 160, 160, 192, 255],
        [255, 255, 192, 192, 160, 160, 128, 128,  96,  64,  64,  96, 128, 128, 160, 160, 192, 192, 255, 255],
        [255, 255, 255, 255, 192, 192, 160, 160, 128,  96,  96, 128, 160, 160, 192, 192, 255, 255, 255, 255],
        [255, 255, 255, 255, 255, 255, 192, 192, 160, 128, 128, 160, 192, 192, 255, 255, 255, 255, 255, 255]
    ]

    hues = [
        [1, 3, 4, 5, 7, 8, 9, 9, 9, 9, 9, 9, 9, 9, 8, 8, 6, 4, 3, 1],
        [1, 2, 3, 3, 5, 6, 7, 7, 8, 8, 8, 8, 9, 9, 8, 7, 5, 3, 2, 1],
        [1, 2, 3, 3, 5, 5, 6, 6, 5, 6, 6, 7, 7, 7, 6, 6, 4, 3, 2, 1],
        [1, 1, 2, 3, 4, 4, 5, 5, 4, 4, 5, 5, 6, 5, 5, 5, 3, 2, 1, 1],
        [1, 1, 2, 2, 4, 4, 4, 4, 4, 4, 5, 5, 4, 4, 4, 4, 2, 2, 1, 1],
        [0, 0, 1, 1, 2, 2, 3, 3, 3, 3, 3, 3, 3, 3, 2, 2, 1, 1, 0, 0],
        [0, 0, 0, 0, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    ]

#    values = [
#        [ 32,   0,   0,   0,   0,   0,   0,   0,   0,  32],
#        [ 64,   0,   0,   0,   0,   0,   0,   0,   0,  64],
#        [ 96,  32,   0,   0,   0,   0,   0,   0,  32,  96],
#        [128,  64,  32,   0,   0,   0,   0,  32,  64, 128],
#        [160,  96,  64,  32,   0,   0,  32,  64,  96, 160],
#        [192, 128,  96,  64,  32,  32,  64,  96, 128, 192],
#        [255, 160, 128,  96,  64,  64,  96, 128, 160, 255],
#        [255, 192, 160, 128,  96,  96, 128, 160, 192, 255],
#        [255, 255, 192, 160, 128, 128, 160, 192, 255, 255],
#        [255, 255, 255, 192, 160, 160, 192, 255, 255, 255]
#    ]

#    hues = [
#        [1, 4, 7, 9, 9, 9, 9, 8, 4, 1],
#        [1, 3, 5, 7, 8, 8, 9, 7, 3, 1],
#        [1, 3, 5, 6, 5, 6, 7, 6, 3, 1],
#        [1, 2, 4, 5, 4, 5, 5, 5, 2, 1],
#        [1, 2, 4, 4, 4, 5, 4, 4, 2, 1],
#        [0, 1, 2, 3, 3, 3, 3, 2, 1, 0],
#        [0, 0, 1, 2, 2, 2, 2, 1, 0, 0],
#        [0, 0, 0, 1, 1, 1, 1, 0, 0, 0],
#        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
#        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
#    ]

    line = [0]*config.LED_ROWS
    matrix = [x[:] for x in [[0]*config.LED_COLS]*config.LED_ROWS]
    leds = [x[:] for x in [[(0, 0, 0)]*config.LED_COLS]*config.LED_ROWS]
    percent = 0

    def __init__(self, ipcon):
        self.okay = False
        self.ipcon = ipcon

        if not config.UID_LED_STRIP_BRICKLET:
            print('Not Configured: LED Strip (required)')
            return

        self.led_strip = LEDStrip(config.UID_LED_STRIP_BRICKLET, self.ipcon)

        try:
            self.led_strip.get_frame_duration()
            print('Found: LED Strip ({0})').format(config.UID_LED_STRIP_BRICKLET)
        except:
            print('Not Found: LED Strip ({0})').format(config.UID_LED_STRIP_BRICKLET)
            return

        self.okay = True

        self.update_frame_rate()
        self.led_strip.register_callback(self.led_strip.CALLBACK_FRAME_RENDERED,
                                         self.frame_rendered)

    def stop_rendering(self):
        if not self.okay:
            return

        self.led_strip.register_callback(self.led_strip.CALLBACK_FRAME_RENDERED,
                                         None)

    def update_frame_rate(self):
        if not self.okay:
            return

        self.led_strip.set_frame_duration(1000.0 / config.FIRE_FRAME_RATE)

    def frame_rendered(self, _):
        self.frame_upload()
        self.frame_prepare_next()

    def frame_upload(self):
        if not self.okay:
            return

        # Reorder LED data into R, G and B channel
        r = []
        g = []
        b = []
        for row in range(config.LED_ROWS):
            col_range = range(config.LED_COLS)
            if row % 2 == 0:
                col_range = reversed(col_range)
            for col in col_range:
                r.append(self.leds[row][col][config.R_INDEX])
                g.append(self.leds[row][col][config.G_INDEX])
                b.append(self.leds[row][col][config.B_INDEX])

        # Make chunks of size 16
        r_chunk = [r[i:i+16] for i in range(0, len(r), 16)]
        g_chunk = [g[i:i+16] for i in range(0, len(g), 16)]
        b_chunk = [b[i:i+16] for i in range(0, len(b), 16)]

        for i in range(len(r_chunk)):
            length = len(r_chunk[i])

            # Fill up chunks with zeros
            r_chunk[i].extend([0]*(16-len(r_chunk[i])))
            g_chunk[i].extend([0]*(16-len(g_chunk[i])))
            b_chunk[i].extend([0]*(16-len(b_chunk[i])))

            try:
                self.led_strip.set_rgb_values(i*16, length, r_chunk[i], g_chunk[i], b_chunk[i])
            except:
                break

    def frame_prepare_next(self):
        def shift_up():
            for y in reversed(range(1, config.LED_COLS)):
                for x in range(config.LED_ROWS):
                    self.matrix[x][y] = self.matrix[x][y-1]

            for x in range(config.LED_ROWS):
                self.matrix[x][0] = self.line[x]

        def generate_line():
            for x in range(config.LED_ROWS):
                self.line[x] = random.randint(min(config.FIRE_RAND_VALUE_START, config.FIRE_RAND_VALUE_END), config.FIRE_RAND_VALUE_END)

        def make_frame():
            def hsv_to_rgb(h, s, v):
                r, g, b = colorsys.hsv_to_rgb(h/255.0, s/255.0, v/255.0)
                return ((int(255*r), int(255*g), int(255*b)))

            for y in reversed(range(1, config.LED_COLS)):
                for x in range(config.LED_ROWS):
                    self.leds[x][config.LED_COLS-1-y] = hsv_to_rgb(self.hues[y][x]*config.FIRE_HUE_FACTOR,
                                                                   255,
                                                                   max(0, (((100.0-self.percent)*self.matrix[x][y] + self.percent*self.matrix[x][y-1])/100.0) - self.values[y][x]))

            for x in range(config.LED_ROWS):
                self.leds[x][config.LED_COLS-1] = hsv_to_rgb(self.hues[0][x]*config.FIRE_HUE_FACTOR,
                                                             255,
                                                             max(0, ((100.0-self.percent)*self.matrix[x][0] + self.percent*self.line[x])/100.0))

        self.percent += 20
        if self.percent >= 100:
            shift_up()
            generate_line()
            self.percent = 0

        make_frame()


if __name__ == "__main__":
    ipcon = IPConnection()
    ipcon.connect(config.HOST, config.PORT)

    fire = Fire(ipcon)
    fire.frame_rendered(0)

    raw_input('Press enter to exit\n') # Use input() in Python 3

    ipcon.disconnect()

File: fire/delphi/Fire.pas
program Fire;

{$ifdef MSWINDOWS}{$apptype CONSOLE}{$endif}
{$ifdef FPC}{$mode OBJFPC}{$H+}{$endif}

uses
  SysUtils, Math, Config, IPConnection, BrickletLEDStrip;

const
  CHUNK_SIZE = 16;
  FIRE_VALUES : array [0..9, 0..19] of byte =
  (
    ( 32,  16,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,  16,  32),
    ( 64,  32,  16,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,  16,  32,  64),
    ( 96,  64,  32,  32,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,  32,  32,  64,  96),
    (128,  96,  64,  64,  32,  32,   0,   0,   0,   0,   0,   0,   0,   0,  32,  32,  64,  64,  96, 128),
    (160, 128,  96,  96,  64,  64,  32,  32,   0,   0,   0,   0,  32,  32,  64,  64,  96,  96, 128, 160),
    (192, 160, 128, 128,  96,  96,  64,  64,  32,   0,   0,  32,  64,  64,  96,  96, 128, 128, 160, 192),
    (255, 192, 160, 160, 128, 128,  96,  96,  64,  32,  32,  64,  96,  96, 128, 128, 160, 160, 192, 255),
    (255, 255, 192, 192, 160, 160, 128, 128,  96,  64,  64,  96, 128, 128, 160, 160, 192, 192, 255, 255),
    (255, 255, 255, 255, 192, 192, 160, 160, 128,  96,  96, 128, 160, 160, 192, 192, 255, 255, 255, 255),
    (255, 255, 255, 255, 255, 255, 192, 192, 160, 128, 128, 160, 192, 192, 255, 255, 255, 255, 255, 255)
  );
  FIRE_HUES : array [0..9, 0..19] of byte =
  (
    (1, 3, 4, 5, 7, 8, 9, 9, 9, 9, 9, 9, 9, 9, 8, 8, 6, 4, 3, 1),
    (1, 2, 3, 3, 5, 6, 7, 7, 8, 8, 8, 8, 9, 9, 8, 7, 5, 3, 2, 1),
    (1, 2, 3, 3, 5, 5, 6, 6, 5, 6, 6, 7, 7, 7, 6, 6, 4, 3, 2, 1),
    (1, 1, 2, 3, 4, 4, 5, 5, 4, 4, 5, 5, 6, 5, 5, 5, 3, 2, 1, 1),
    (1, 1, 2, 2, 4, 4, 4, 4, 4, 4, 5, 5, 4, 4, 4, 4, 2, 2, 1, 1),
    (0, 0, 1, 1, 2, 2, 3, 3, 3, 3, 3, 3, 3, 3, 2, 2, 1, 1, 0, 0),
    (0, 0, 0, 0, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 0, 0, 0, 0),
    (0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0),
    (0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0),
    (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)
  );

type
  TFire = class
  private
    ipcon: TIPConnection;
    ledStrip: TBrickletLEDStrip;
    line: array [0..(LED_ROWS - 1)] of integer;
    matrix: array [0..(LED_ROWS - 1), 0..(LED_COLS - 1)] of integer;
    leds: array [0..(LED_ROWS - 1), 0..(LED_COLS - 1), 0..2] of byte;
    percent: integer;
  public
    constructor Create;
    procedure HSV2RGB(const h, s, v: byte; out r, g, b: byte);
    function Interpolate2(const x, y: integer): integer;
    function Interpolate1(const x: integer): integer;
    procedure FrameRenderedCB(sender: TBrickletLEDStrip; const length: word);
    procedure FrameUpload;
    procedure FramePrepareNext;
    procedure Execute;
  end;

var
  e: TFire;

constructor TFire.Create;
begin

  inherited Create;
end;

{ Based on http://web.mit.edu/storborg/Public/hsvtorgb.c }
procedure TFire.HSV2RGB(const h, s, v: byte; out r, g, b: byte);
var region, fpart, p, q, t: byte;
begin
  if (s = 0) then begin
    { Color is grayscale }
    r := v;
    g := v;
    b := v;
  end
  else begin
    { Make hue 0-5 }
    region := Floor(h / 43);

    { Find remainder part, make it from 0-255 }
    fpart := (h - (region * 43)) * 6;

    { Calculate temp vars, doing integer multiplication }
    p := (v * (255 - s)) shr 8;
    q := (v * (255 - ((s * fpart) shr 8))) shr 8;
    t := (v * (255 - ((s * (255 - fpart)) shr 8))) shr 8;

    { Assign temp vars based on color cone region }
    case (region) of
    0: begin   r := v; g := t; b := p; end;
    1: begin   r := q; g := v; b := p; end;
    2: begin   r := p; g := v; b := t; end;
    3: begin   r := p; g := q; b := v; end;
    4: begin   r := t; g := p; b := v; end;
    else begin r := v; g := p; b := q; end;
    end;
  end;
end;

function TFire.Interpolate2(const x, y: integer): integer;
var p0, p1, m0, m1: integer;
begin
  p0 := 100 - percent;
  p1 := percent;
  m0 := matrix[x, y];
  m1 := matrix[x, y - 1];

  result := Floor((p0 * m0 + p1 * m1) / 100);
end;

function TFire.Interpolate1(const x: integer): integer;
var p0, p1, m0, m1: integer;
begin
  p0 := 100 - percent;
  p1 := percent;
  m0 := matrix[x, 0];
  m1 := line[x];

  result := Floor((p0*m0 + p1*m1) / 100);
end;

procedure TFire.FrameRenderedCB(sender: TBrickletLEDStrip; const length: word);
begin
  FrameUpload;
  FramePrepareNext;
end;

procedure TFire.FrameUpload;
var r, g, b: array [0..(LED_ROWS * LED_COLS - 1)] of byte;
    rChunk, gChunk, bChunk: array [0..(CHUNK_SIZE - 1)] of byte;
    row, col, colBegin, colEnd, colStep, i, k: integer;
begin
  { Reorder LED data into R, G and B channel }
  i := 0;
  for row := 0 to (LED_ROWS - 1) do begin
    if (row mod 2 = 0) then begin
      colBegin := LED_COLS - 1;
      colEnd := -1;
      colStep := -1;
    end
    else begin
      colBegin := 0;
      colEnd := LED_COLS;
      colStep := 1;
    end;

    col := colBegin;
    while (col <> colEnd) do begin
      r[i] := leds[row, col, R_INDEX];
      g[i] := leds[row, col, G_INDEX];
      b[i] := leds[row, col, B_INDEX];
      i += 1;
      col += colStep;
    end;
  end;

  { Make chunks of size 16 }
  i := 0;
  while (i < LED_ROWS * LED_COLS) do begin
    k := 0;
    while ((k < CHUNK_SIZE) and (i + k < LED_ROWS * LED_COLS)) do begin
      rChunk[k] := r[i + k];
      gChunk[k] := g[i + k];
      bChunk[k] := b[i + k];
      k += 1;
    end;

    try
      ledStrip.SetRGBValues(i, k, rChunk, gChunk, bChunk);
    except
      exit;
    end;

    i += CHUNK_SIZE;
  end;
end;

procedure TFire.FramePrepareNext;
var x, y, start: integer; r, g, b: byte;
begin
  percent += 20;
  if (percent >= 100) then begin
    { Shift up }
    for y := (LED_COLS - 1) downto 1 do begin
      for x := 0 to (LED_ROWS - 1) do begin
        matrix[x, y] := matrix[x, y - 1]
      end;
    end;
    for x := 0 to (LED_ROWS - 1) do begin
      matrix[x, 0] := line[x]
    end;

    { Generate_line }
    start := Min(FIRE_RAND_VALUE_START, FIRE_RAND_VALUE_END);
    for x := 0 to (LED_ROWS - 1) do begin
      line[x] := start + Random(FIRE_RAND_VALUE_END - start);
    end;

    percent := 0;
  end;

  { Make frame }
  for y := (LED_COLS - 1) downto 1 do begin
    for x := 0 to (LED_ROWS - 1) do begin
      HSV2RGB(Floor(FIRE_HUES[y][x] * FIRE_HUE_FACTOR), 255,
              Max(0, Interpolate2(x, y) - FIRE_VALUES[y, x]), r, g, b);
      leds[x, LED_COLS-1-y, 0] := r;
      leds[x, LED_COLS-1-y, 1] := g;
      leds[x, LED_COLS-1-y, 2] := b;
    end;
  end;
  for x := 0 to (LED_ROWS - 1) do begin
    HSV2RGB(Floor(FIRE_HUES[0][x] * FIRE_HUE_FACTOR), 255,
            Max(0, Interpolate1(x)), r, g, b);
    leds[x, LED_COLS-1, 0] := r;
    leds[x, LED_COLS-1, 1] := g;
    leds[x, LED_COLS-1, 2] := b;
  end;
end;

procedure TFire.Execute;
begin
  { Create IP Connection and connect it }
  ipcon := TIPConnection.Create;
  ipcon.Connect(HOST, PORT);

  { Call a getter to check that the Bricklet is avialable }
  ledStrip := TBrickletLEDStrip.Create(UID_LED_STRIP_BRICKLET, ipcon);

  try
    ledStrip.GetFrameDuration;
    WriteLn(Format('Found: LED Strip %s', [UID_LED_STRIP_BRICKLET]));
  except
    WriteLn(Format('Not Found: LED Strip %s', [UID_LED_STRIP_BRICKLET]));
    exit;
  end;

  ledStrip.SetFrameDuration(Floor(1000 / FIRE_FRAME_RATE));

  { Register frame rendered callback to function FrameRenderedCB }
  ledStrip.OnFrameRendered := {$ifdef FPC}@{$endif}FrameRenderedCB;

  { Start rendering }
  FrameRenderedCB(ledStrip, 0);

  WriteLn('Press key to exit');
  ReadLn;
  ipcon.Destroy; { Calls ipcon.Disconnect internally }
end;

begin
  e := TFire.Create;
  e.Execute;
  e.Destroy;
end.
